include/flux/op/cartesian_product.hpp

```diff
@@ -216,6 +231,33 @@ struct sequence_traits<detail::cartesian_product_adaptor<Bases...>>
         }(std::index_sequence_for<Bases...>{});
     }
 
+    template <std::size_t I, typename Self, typename Function,
+              typename... PartialElements>
+    static constexpr void for_each_while_impl(Self& self,
+                                              bool& keep_going,
+                                              cursor_t<Self>& cur,
+                                              Function&& func,
+                                              PartialElements&&... partial_elements)
+    {
+        // We need to iterate right to left.
+        if constexpr (I == sizeof...(Bases) - 1) {
+            std::get<I>(cur) = flux::for_each_while(std::get<I>(self.bases_),
+                [&](auto&& elem) {
+                    keep_going = std::invoke(func,
+                        element_t<Self>(FLUX_FWD(partial_elements)..., FLUX_FWD(elem)));
+                    return keep_going;
+                });
+        } else {
+            std::get<I>(cur) = flux::for_each_while(std::get<I>(self.bases_),
+                [&](auto&& elem) {
+                    for_each_while_impl<I+1>(
+                        self, keep_going, cur,
+                        func, FLUX_FWD(partial_elements)..., FLUX_FWD(elem));
+                    return keep_going;
+                });
+        }
+    }
```

Comment:
I love this implementation! It's super-clean, and nice to be able to avoid the `partial_cursor` work as well

Reply:
Thanks! This is definitely much better, thanks for pushing me in the right direction.